pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    @Override
+    public void initialize(ServiceConfiguration conf, ConfigurationCacheService configCache) throws IOException {
+        String prefix = (String) conf.getProperty(CONF_TOKEN_SETTING_PREFIX);
+        if (null == prefix) {
+            prefix = "";
+        }
+        String confTokenAuthClaimSettingName = prefix + CONF_TOKEN_AUTH_CLAIM;
+        if (conf.getProperty(confTokenAuthClaimSettingName) != null
+                && StringUtils.isNotBlank((String) conf.getProperty(confTokenAuthClaimSettingName))) {
+            this.roleClaim = (String) conf.getProperty(confTokenAuthClaimSettingName);
+        }
+
+        super.initialize(conf, configCache);
+    }
+
+    private List<String> getRoles(AuthenticationDataSource authData) {
+        String token = null;
+
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            token = authData.getCommandData();
+            if (StringUtils.isBlank(token)) {
+                return Collections.emptyList();
+            }
+        } else if (authData.hasDataFromHttp()) {
+            // The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                return Collections.emptyList();
+            }
+
+            // Remove prefix
+            token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+        }
+
+        if (token == null)
+            return Collections.emptyList();
+
+        String[] splitToken = token.split("\\.");
+        String unsignedToken = splitToken[0] + "." + splitToken[1] + ".";
+
+        Jwt<?, Claims> jwt = parser.parseClaimsJwt(unsignedToken);
+        try {
+            Collections.singletonList(jwt.getBody().get(roleClaim, String.class));
+        } catch (RequiredTypeException requiredTypeException) {
+            List list = jwt.getBody().get(roleClaim, List.class);
+            if (list != null) {
+                return list;
+            }
+        }
+
+        return Collections.emptyList();
+    }
+
+    public CompletableFuture<Boolean> authorize(AuthenticationDataSource authenticationData, Function<String, CompletableFuture<Boolean>> authorizeFunc) {
+        List<String> roles = getRoles(authenticationData);
+        List<CompletableFuture<Boolean>> futures = new ArrayList<>(roles.size());
+        roles.forEach(r -> futures.add(authorizeFunc.apply(r)));
+        CompletableFuture[] cfs = futures.toArray(new CompletableFuture[futures.size()]);
+        return CompletableFuture.allOf(cfs)
+                .thenApply(ignored -> futures.stream().map(CompletableFuture::join).collect(Collectors.toList()))
+                .thenApply(list -> list.stream().anyMatch(result -> result));
```

Comment:
I think we can improve the implementation here since we don't need to wait for all furtures to be complete right? If one `authorizeFunc` completes and responds for `true`, then we can complete the result future.